### PR TITLE
STY: Apply pre-commit hooks to codebase

### PR DESCRIPTION
**Pull request by matthewfeickert**
## PR Summary

This is a follow up PR to PR #21583 based on discussions in PR #22777 (c.f. https://github.com/matplotlib/matplotlib/pull/22777#issuecomment-1093039541) and [on Gitter](https://gitter.im/matplotlib/matplotlib?at=625001120466b352a458d224). This touches a lot of files so I'll walk through the story here, which should help motivate turning on pre-commit.ci.

1. The pre-commit hooks were added in PR #21583 (November 2021) with a pre-commit.ci 

https://github.com/matplotlib/matplotlib/blob/a436253d1d672cd46b0da93acdd868d792f2dbb8/.pre-commit-config.yaml#L3

though as pre-commit.ci wasn't turned on, there were no pre-commit hook updates coming in on that schedule and so I've run `pre-commit autoupdate` to update the `pre-commit-hooks` and `flake8` ids.

2. With those updated, I then ran the failing (c.f. https://github.com/matplotlib/matplotlib/pull/22777#issuecomment-1093039541) `pre-commit` hooks and applied their necessary revisions as individual commits. So those are the style commits for `end-of-file-fixer`, `trailing-whitespace`, and `check-docstring-first`. These were also done as individual commits so that if in review any of these are viewed as not acceptable they can be dropped easily with a rebase.

3. As the following commits touch many files, this is a noisy first time run of the hooks. To help limit this, add a [`.git-blame-ignore-revs` file](https://git-scm.com/docs/git-blame#Documentation/git-blame.txt---ignore-revs-fileltfilegt) that will [automatically ignore commits in the blame view on GitHub](https://docs.github.com/en/repositories/working-with-files/using-files/viewing-a-file#ignore-commits-in-the-blame-view) (this file is required to be in the top level of the repo). This can also be done locally with `git blame` by running

```
git blame --ignore-revs-file .git-blame-ignore-revs
```

Example on my fork:
[example_ignore](https://github.com/matthewfeickert/matplotlib/blame/mnt/apply-pre-commit-to-all-files/lib/matplotlib/_constrained_layout.py)

**Note to reviewer**: As this file contains commits it is sensitive to rebases, and so please make sure that these commits are correct if anyone rebases this during review.

This PR should probably be merged in advance of PR #22777 which can be trivially rebased and would be a good time to turn pre-commit.ci on.

`pre-commit` now passes the full repo:

```console
$ pre-commit run --all-files
check for added large files..............................................Passed
check docstring is first.................................................Passed
fix end of files.........................................................Passed
mixed line ending........................................................Passed
trim trailing whitespace.................................................Passed
flake8...................................................................Passed
```

## PR Checklist

**Tests and Styling**
- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [n/a] New features are documented, with examples if plot related.
- [n/a] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [n/a] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

File: .git-blame-ignore-revs [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# style: end-of-file-fixer pre-commit hook
+b372d66ab9e9aa30573f240d87a6ea92474cce75
+
+# style: trailing-whitespace pre-commit hook
+0f0a977d8aaba773570dff8f828c358df22203aa
+
+# style: check-docstring-first pre-commit hook
+1a36dc83af39159749ce5857eca6ddacd345e6bb
```
Comment (matthewfeickert) [inline .git-blame-ignore-revs:8]: Note to reviewer: As this file contains commits it is sensitive to rebases, and so please make sure that these commits are correct if anyone rebases this during review.

File: doc/devel/development_setup.rst [hunk 2]
```diff
@@ -67,9 +67,18 @@ These will automatically check flake8 and other style issues when you run
...
     pre-commit install
 
+The hooks can also be run manually. All the hooks can be run, in order as
+listed in ``.pre-commit-config.yaml``, against the full codebase with ::
+
+    pre-commit run --all-files
+
+To run a particular hook manually, run ``pre-commit run`` with the hook id ::
+
+    pre-commit run <hook id> --all-files
```
Comment (matthewfeickert) [inline doc/devel/development_setup.rst:80]: You can of course target particular files with `--file` (e.g. `pre-commit run --file README.rst`) but given that is already done on the files that you're attempting to commit when you have pre-commit installed locally this didn't really seem important to note.

From `pre-commit run --help`:
```
  --files [FILES ...]   Specific filenames to run hooks on.
```

**Comment by matthewfeickert** [comment]
To address a [question from Gitter](https://gitter.im/matplotlib/matplotlib?at=6250b4dbddcba117a25fbf96):

In PR #17143 `reviewdog` was added. `reviewdog` currently appears to run linting for only `flake8` (please correct me if I'm wrong). As to how `pre-commit` differs, it applies all the `pre-commit` hooks (including `flake8`) and will quarterly (as defined in the config)

https://github.com/matplotlib/matplotlib/blob/a436253d1d672cd46b0da93acdd868d792f2dbb8/.pre-commit-config.yaml#L3

open a PR to automatically update the hooks that have new releases. It also runs all of the pre-commit hooks faster then `reviewdog` (though this really doesn't matter at all when you're comparing against how long the rest of the CI takes).

**Comment by matthewfeickert** [comment]
I also enabled pre-commit.ci on my fork to do a demo of what this looks like. c.f. https://github.com/matthewfeickert/matplotlib/pull/9 (which is mostly just the following)

[my_fork](https://results.pre-commit.ci/run/github/196771788/1649460753.hkVpcXgaQkqBK3Su-cveqg)

**Comment by oscargus** [comment]
I'm not sure that we should require pre-commit hooks (as I take is an indirect suggestion here). While I tend to use it, there are setups where it is not really convenient. For example, I have different shells on Windows, one for running git (https://gitforwindows.org/) and one for Python (conda). While I can of course unify them (being too lazy so far, and doing much development on Linux), this is a bit of extra work and nothing I think should be an obstacle for contributing.

Also, I am not sure how this works when doing commits from graphical interfaces (which for some reason seems to be a popular way of using git...).

Running the hooks on all files is as you say a one-time effort. IMHO I do not think we need to use the `.git-blame-ignore-revs` here, but just go for it (if we should do it). All newly committed code will have these checks anyway and if someone touches some code and have white-space removal in the editor, it will be enforced anyway. With that said, I am not against cleaning the code in one go, but think the `.git-blame-ignore-revs` is really not required.

To me, pre-commit hooks primarily serve the purpose of not having to run the code quality tests locally and get warnings if something is not right, so I am not a big fan of the hooks actually changing stuff. Better to "learn" how it should be done.

**Comment by matthewfeickert** [comment]
> I'm not sure that we should require pre-commit hooks (as I take is an indirect suggestion here).

This doesn't require developers to use them. Using pre-commit.ci is just going to find things that would later require a developer to go fix to and alert the PR contributor and the reviewers to them. 

> While I tend to use it, there are setups where it is not really convenient. For example, I have different shells on Windows, one for running git (https://gitforwindows.org/) and one for Python (conda). While I can of course unify them (being too lazy so far, and doing much development on Linux), this is a bit of extra work and nothing I think should be an obstacle for contributing.

I don't think I understand. :? What is the obstacle? Are you saying that `pre-commit` doesn't work on your Windows setup? Can you elaborate?

> Also, I am not sure how this works when doing commits from graphical interfaces (which for some reason seems to be a popular way of using git...).

That is an interesting point that I've never thought about before. I would be a bit surprised (maybe naively?) to find out that popular Git GUIs have no plugin support though, and so I would (definitely naively) hope things "just work".

> Running the hooks on all files is as you say a one-time effort. IMHO I do not think we need to use the `.git-blame-ignore-revs` here, but just go for it (if we should do it). All newly committed code will have these checks anyway and if someone touches some code and have white-space removal in the editor, it will be enforced anyway. With that said, I am not against cleaning the code in one go, but think the `.git-blame-ignore-revs` is really not required.

It is certainly not required, but the diff is pretty ugly and meaningless, so this was more of a "making it less painful" effort. So I'm quite happy to remove this, but the blame avoidance seems almost custom designed for these huge diffs situations.

> To me, pre-commit hooks primarily serve the purpose of not having to run the code quality tests locally and get warnings if something is not right, so I am not a big fan of the hooks actually changing stuff. Better to "learn" how it should be done.

I take a different view on their purpose (automate the boring stuff that would need to get human changed anyway), but my opinion on that isn't the relevant thing here. The hooks only run locally if the developer has them installed locally (which is totally optional) and if a developer pushes code that fails pre-commit.ci then as automatic pushing back fixes is disabled

https://github.com/matplotlib/matplotlib/blob/a436253d1d672cd46b0da93acdd868d792f2dbb8/.pre-commit-config.yaml#L2

they can speed up the process and avoid having a reviewer have to ask for them for changes by just commenting "pre-commit.ci autofix". Nothing is happening automatically unless a developer opts in by using pre-commit locally or by asking pre-commit.ci to fix things for them. (I think I am probably telling you things that you already know though, so apologies here. :grimacing: Can you elaborate/clarify your points if I've misunderstood?)

I'll note also that this PR is simply enforcing the pre-commit hooks that were added in PR #21583, and if the dev team isn't happy with that hook selection I'm certainly happy to fix that here to whatever is desired. :+1:

File: .git-blame-ignore-revs (hunk 1 above)
Comment (jklymak) [reply .git-blame-ignore-revs:8]: This is great that this is possible.  Really helpful for big cosmetic changes like this.

**Review by jklymak — APPROVED** [review]
With the git-blame-ignore-revs, I think these cosmetic changes are fine to get the recommit hooks to pass.

**Comment by oscargus** [comment]
> I don't think I understand. :? What is the obstacle? Are you saying that `pre-commit` doesn't work on your Windows setup? Can you elaborate?

The problem is that the git (for Windows) bash shell and the (mini)conda shell (Anaconda prompt) don't have the same Python environment, so if I run git in the conda shell I have everything I need for the pre-commit, but not the nice aspects of the git bash shell (like branch name and tab-completion), and if I use the git shell, I do not have the conda environment.

As I said, this can be sorted out with a bit of effort and is just a convenience thing really, but I would assume it is a similar situation for the graphical clients as conda typically doesn't set up any environment variables except for it's own shell.

(I think I have three different Python installs on my Windows machine: conda, that I use for most things, Windows Python that ends up in the path, and the one provided with the MS dev-prompt for building certain stuff. So while I can have all three synced, it is not very practical...)

File: lib/matplotlib/backend_tools.py [hunk 3]
```diff
@@ -989,12 +989,12 @@ def trigger(self, *args, **kwargs):
                  'help': ToolHelpBase,
                  'copy': ToolCopyToClipboardBase,
                  }
-"""Default tools"""
+# Default tools
```
Comment (greglucas) [inline lib/matplotlib/backend_tools.py:992]: Just delete this comment and the one below, the comment is the same as the variable name :)

**Review by greglucas — APPROVED** [review]
I'm actually surprised there is this little churn in the files, it is almost all trailing whitespace.

I also prefer having the `.git-blame-ignore-revs` for the whitespace changes because what I'm really interested in with the blame is looking for who updated the actual code piece and I don't want to ping @matthewfeickert to fix something he only did a whitespace update on :)

File: lib/matplotlib/backend_tools.py (hunk 3 above)
Comment (matthewfeickert) [reply lib/matplotlib/backend_tools.py]: Done in rebase I just pushed. :+1:

File: .git-blame-ignore-revs (hunk 1 above)
Comment (matthewfeickert) [reply .git-blame-ignore-revs]: I rebased, but I also updated the commits.

```console
$ git log --graph --oneline --decorate --all | head -n 10
* 02fc0611cf (HEAD -> mnt/apply-pre-commit-to-all-files, origin/mnt/apply-pre-commit-to-all-files) DOC: Add pre-commit hook run dev instructions
* f3aa826d83 DEV: Add .git-blame-ignore-revs to ignore style in blame view
* 66c12f2778 STY: Apply check-docstring-first pre-commit hook
* 33516082c5 STY: Apply trailing-whitespace pre-commit hook
* 94d985a744 STY: Apply end-of-file-fixer pre-commit hook
* 7e04925fcd MNT: Update pre-commit hooks to latest versions
| * 3dd6049160 (mnt/add-codespell-to-pre-commit-hooks) FIX: Fix typo caught by codespell pre-commit hook
| * 19987645ec DEV: Add codespell to pre-commit hooks
|/  
*   19992285cb (upstream/main, origin/main, main) Merge pull request #22807 from anntzer/qd
```

**Comment by matthewfeickert** [comment]
Just showing that things still pass after rebase:

```console
$ pre-commit run --all-files
check for added large files..............................................Passed
check docstring is first.................................................Passed
fix end of files.........................................................Passed
mixed line ending........................................................Passed
trim trailing whitespace.................................................Passed
flake8...................................................................Passed
```

File: lib/matplotlib/backend_tools.py [hunk 4]
```diff
@@ -989,12 +989,10 @@ def trigger(self, *args, **kwargs):
                  'help': ToolHelpBase,
                  'copy': ToolCopyToClipboardBase,
                  }
-"""Default tools"""
```
Comment (oscargus) [inline lib/matplotlib/backend_tools.py:992]: Why are these removed? Not sure if they are actually rendered, but those are doc-strings for the attributes.

**Comment by oscargus** [comment]
> I don't want to ping

That is the general reason, yes. However, looking at the changes here: do you really expect that you will have to ping for any of these changes? They are primarily in non-code and the very few that are in the code are quite non-critical.

**Comment by greglucas** [comment]
I agree that this specific change isn't too invasive and wouldn't be a big issue either way for the ignore commits from my perspective. However, I see this as setting a good precedent for potential future updates dropping big style changes and makes it easy to see how to do it properly by just adding the commit hash to this file.

File: lib/matplotlib/backend_tools.py (hunk 4 above)
Comment (matthewfeickert) [reply lib/matplotlib/backend_tools.py:992]: They were removed given @greglucas review comment here: https://github.com/matplotlib/matplotlib/pull/22809#discussion_r846630906

**Comment by matthewfeickert** [comment]
> However, I see this as setting a good precedent for potential future updates dropping big style changes and makes it easy to see how to do it properly by just adding the commit hash to this file.

Yeah. :+1: I got the idea from @henryiii using it in https://github.com/scikit-build/scikit-build/pull/685 where they use Black. I know that Matplotlib doesn't use any formatters like that, but if there was ever a big style revision it could be done with a diminished disruption with respect to using the Git history.

**Review by tacaswell — APPROVED** [review]
I have a slight preference that the LICENCES files not be touched / be ignored by these tools but not a big enough concern to block over it.

**Comment by matthewfeickert** [comment]
> I have a slight preference that the LICENCES files not be touched / be ignored by these tools but not a big enough concern to block over it.

@tacaswell That can easily happen. They can just get added to the ignore list

https://github.com/matplotlib/matplotlib/blob/c33557d120eefe3148ebfcf2e758ff2357966000/.pre-commit-config.yaml#L4-L12

I'll do that, rebase to get the right commits, and then request re-review.

File: .git-blame-ignore-revs [hunk 5]
```diff
@@ -0,0 +1,8 @@
+# style: end-of-file-fixer pre-commit hook
+c1a33a481b9c2df605bcb9bef9c19fe65c3dac21
+
+# style: trailing-whitespace pre-commit hook
+213061c0804530d04bbbd5c259f10dc8504e5b2b
+
+# style: check-docstring-first pre-commit hook
+046533797725293dfc2a6edb9f536b25f08aa636
```
Comment (matthewfeickert) [inline .git-blame-ignore-revs:8]: After this last rebase:

```
$ git log --graph --oneline --decorate --all | head -n 7
* 48ade7d904 (HEAD -> mnt/apply-pre-commit-to-all-files, origin/mnt/apply-pre-commit-to-all-files) DOC: Add pre-commit hook run dev instructions
* 4c003775bb DEV: Add .git-blame-ignore-revs to ignore style in blame view
* 0465337977 STY: Apply check-docstring-first pre-commit hook
* 213061c080 STY: Apply trailing-whitespace pre-commit hook
* c1a33a481b STY: Apply end-of-file-fixer pre-commit hook
* 787bdf08e5 MNT: Update pre-commit hooks to latest versions
* 95038d96a4 DEV: Exclude LICENSE files from pre-commit
```

and

```console
$ git grep "c1a33a481b\|213061c080\|0465337977"
.git-blame-ignore-revs:c1a33a481b9c2df605bcb9bef9c19fe65c3dac21
.git-blame-ignore-revs:213061c0804530d04bbbd5c259f10dc8504e5b2b
.git-blame-ignore-revs:046533797725293dfc2a6edb9f536b25f08aa636
```

so these are correct. :+1:

**Comment by matthewfeickert** [comment]
@tacaswell The `LICENSE/` dir is properly excluded now:

```console
$ pre-commit run --all-files
check for added large files..............................................Passed
check docstring is first.................................................Passed
fix end of files.........................................................Passed
mixed line ending........................................................Passed
trim trailing whitespace.................................................Passed
flake8...................................................................Passed
$ git diff upstream/main -- LICENSE/  # No diff
$
```

File: doc/api/lines_api.rst [hunk 6]
```diff
@@ -26,4 +26,4 @@ Functions
    :template: autosummary.rst
 
    segment_hits
-   
\ No newline at end of file
+
```
Comment (QuLogic) [inline doc/api/lines_api.rst]: This seems to be adding a line with trailing whitespace, only to have it removed in a later remove-trailing-whitespace commit?

Comment (matthewfeickert) [reply doc/api/lines_api.rst]: @QuLogic I'm running the pre-commit hooks sequentially and applying their effects so that if someone did want to look at what was actually done (even though these will be (~mostly) ignored in `git blame`) they can see the effects instead of just getting a single commit that is just the result of applying all of the changes for

```console
$ pre-commit run --all-files
```

to pass. So sequentially that is `end-of-file-fixer` which enforces a newline followed by `trailing-whitespace` which removes the trailing whitespace that existed on the line that previously didn't have a newline following.

```console
$ git show c1a33a481b -- doc/api/lines_api.rst
commit c1a33a481b9c2df605bcb9bef9c19fe65c3dac21
Author: Matthew Feickert <matthew.feickert@cern.ch>
Date:   Fri Apr 8 16:02:24 2022 -0500

    STY: Apply end-of-file-fixer pre-commit hook

diff --git a/doc/api/lines_api.rst b/doc/api/lines_api.rst
index 808df726d1..44e0ec1cab 100644
--- a/doc/api/lines_api.rst
+++ b/doc/api/lines_api.rst
@@ -26,4 +26,4 @@ Functions
    :template: autosummary.rst

    segment_hits
-
\ No newline at end of file
+
$ git show 213061c080 -- doc/api/lines_api.rst
commit 213061c0804530d04bbbd5c259f10dc8504e5b2b
Author: Matthew Feickert <matthew.feickert@cern.ch>
Date:   Fri Apr 8 16:05:04 2022 -0500

    STY: Apply trailing-whitespace pre-commit hook

diff --git a/doc/api/lines_api.rst b/doc/api/lines_api.rst
index 44e0ec1cab..4cde67c7e6 100644
--- a/doc/api/lines_api.rst
+++ b/doc/api/lines_api.rst
@@ -26,4 +26,3 @@ Functions
    :template: autosummary.rst

    segment_hits
-
```

Their composite is then the last line in the file having the whitespace stripped and a newline added, seen here.

Is this what your question was on? Or did I misunderstand you?

**Comment by oscargus** [comment]
> setting a good precedent

Fair enough!

File: doc/api/lines_api.rst (hunk 6 above)
Comment (matthewfeickert) [reply doc/api/lines_api.rst]: Yeah it unfortunately doesn't show up well in the diff. Thanks for clarifying! :)

File: lib/matplotlib/backend_tools.py (hunk 4 above)
Comment (matthewfeickert) [reply lib/matplotlib/backend_tools.py:992]: @oscargus I'll let you and the other members of the dev team make a call here, but the fact that this does show up in the [the current docs](https://matplotlib.org/devdocs/api/backend_tools_api.html?highlight=default_tools#matplotlib.backend_tools.default_tools) looks unintentional to me. At least I don't see their value here in the docs, but I'm also obviously not an expert in this section of the API.

not_intentional

These docstring comments were added in 2015 back in PR #3652 by @fariza. So while I sure hope no one ever asks me to explain code that I wrote over 5 years ago :grimacing: maybe they can take a look and give their thoughts.

@greglucas as you suggested removal maybe you can also give your thoughts here.

Comment (oscargus) [reply lib/matplotlib/backend_tools.py:992]: I agree that it is not the most relevant section of the MPL documentation, so I do not see any real issue with letting it go, especially `default_tools`. Just wanted to point out the consequence.

Comment (oscargus) [reply lib/matplotlib/backend_tools.py:992]: Can you please clarify if there is a pre-commit hook that made you turn the doc-string into a comment? (As I assume that the purpose was not to get rid of the documentation.)

Would be good to know in case we at some stage start to enforce this. (Which we probably shouldn't then.)

Comment (matthewfeickert) [reply lib/matplotlib/backend_tools.py:992]: > Just wanted to point out the consequence.

Sure. I totally get that, and I truly appreciate you thinking about this PR so carefully. So thank you for being a great reviewer! :)

Comment (matthewfeickert) [reply lib/matplotlib/backend_tools.py:992]: > Can you please clarify if there is a pre-commit hook that made you turn the doc-string into a comment? (As I assume that the purpose was not to get rid of the documentation.)

That's part of what the [`check-docstring-first` hook](https://github.com/pre-commit/pre-commit-hooks/tree/c7a7ffb9a71a6859cf588f967bb8a208d5871d53#check-docstring-first) does

```
$ git checkout upstream/main -- lib/matplotlib/backend_tools.py
$ pre-commit run check-docstring-first --all-files
check docstring is first.................................................Failed
- hook id: check-docstring-first
- exit code: 1

lib/matplotlib/backend_tools.py:992: Multiple module docstrings (first docstring on line 1).

```

> Would be good to know in case we at some stage start to enforce this. (Which we probably shouldn't then.)

I think the use cases for having a real use for multiple module level docstrings is quite rare. At that point might as well just ignore that particular file for this particular hook.

I'll note again though that I don't have any strong feelings on `check-docstring-first` being used or not used in general &mdash; it is just one of the pre-commit hooks added in PR #21583.

Comment (fariza) [reply lib/matplotlib/backend_tools.py:992]: That was added just following the general directions that all public variables should be documented. There was no specific intent to it.
And specially since the previous entry shows the same list being added, it is not really needed.

Comment (matthewfeickert) [reply lib/matplotlib/backend_tools.py:992]: Thanks @fariza!

Comment (QuLogic) [reply lib/matplotlib/backend_tools.py:992]: I think the question is not whether we should remove these strings (because they do seem to not provide much more information anyway), but whether pre-commit will preclude us from ever documenting constants like this, and whether that is ever going to be something that we need to do.

Comment (matthewfeickert) [reply lib/matplotlib/backend_tools.py:992]: The good news is that whatever you decide the impact is quite low.

* If you decide to keep this hook and then later decide that it is impeding you, then you add the file that is failing to `check-docstring-first`'s specific `exclude`. Or you delete `check-docstring-first`
* If you don't want `check-docstring-first` then we delete it from pre-commit and optionally drop these changes.

So maybe the most relevant question is: can think of any examples where you would want multiple module level docstrings? The hooks all pass so you don't have any else in the codebase. If in the entire project these were the only ones and they were added in 2015 then it seems a safe bet this isn't going to change in the near future unless you already have examples in mind.

Comment (henryiii) [reply lib/matplotlib/backend_tools.py:992]: FYI, check-docstring-first had a long-standing issue for these false positives, see https://github.com/pre-commit/pre-commit-hooks/issues/159. It can be turned off per-file, and does catch the common case where imports sneak above the docstring. Or it can be dropped, either way.

Comment (QuLogic) [reply lib/matplotlib/backend_tools.py:992]: I think this affects excess docstrings in examples far more than variable docstrings in the actual library, so I think I'm net positive on keeping it.

Comment (matthewfeickert) [reply lib/matplotlib/backend_tools.py:992]: > so I think I'm net positive on keeping it.

(Sorry, long week and my brain is a bit fuzzy.) Positive on keeping the `check-docstring-first` hook, correct?

Comment (greglucas) [reply lib/matplotlib/backend_tools.py:992]: I'm pretty indifferent on this (as it sounds like most of us are!). I'd lean towards sticking with what you have currently (removing the docstring/comment [and because of that removing the sphinx generated documentation] and leaving the pre-commit check in place).

Comment (matthewfeickert) [reply lib/matplotlib/backend_tools.py:992]: @QuLogic @oscargus @jklymak @greglucas @tacaswell As it seems that there are no strong feelings against keeping the `check-docstring-first` hook and some positive leanings towards it, does anyone have any further questions on this PR?

To keep the noise down on this I've stopped rebasing it against `main` but I have a branch on my fork where I've continued to do so and `pre-commit run --all-files` is still passing with this rebased on `main`'s `HEAD`.

**Comment by tacaswell** [comment]
Thanks @matthewfeickert !
